sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala

```diff
@@ -117,7 +123,10 @@ class CSVInferSchema(val options: CSVOptions) extends Serializable {
         case LongType => tryParseLong(field)
         case _: DecimalType => tryParseDecimal(field)
         case DoubleType => tryParseDouble(field)
+        case DateType => tryParseDateTime(field)
+        case TimestampNTZType if options.inferDate => tryParseDateTime(field)
```

Comment:
shall we call `tryParseTimestampNTZ`? The type so far is timestamp, and inferring date is useless as we need to promote it to timestamp anyway.

Reply:
Our expected behavior is that in a column with `TimestampType` entries and then `DateType` entries, the column will be inferred as `TimestampType`. 

Here, `tryParseTimestampNTZ` and `tryParseTimestamp` will not be able to parse the `DateType` entries that show up later in the column and the column will be promoted to string type. So we must use `tryParseDateTime` which will give a chance for the date to be parsed.